app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationManager.kt

```diff
@@ -0,0 +1,311 @@
+import at.connyduck.calladapter.networkresult.onSuccess
+import com.google.android.material.snackbar.Snackbar
+import com.keylesspalace.tusky.R
+import com.keylesspalace.tusky.components.login.LoginActivity
+import com.keylesspalace.tusky.db.AccountEntity
+import com.keylesspalace.tusky.entity.Notification
+import com.keylesspalace.tusky.entity.NotificationSubscribeResult
+import com.keylesspalace.tusky.network.MastodonApi
+import com.keylesspalace.tusky.util.CryptoUtil
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.unifiedpush.android.connector.UnifiedPush
+import retrofit2.HttpException
+import javax.inject.Inject
+
+// TODO architecture-wise: the NotificationHelper should probably be a NotificationManager which either uses
+//   pull or push notifications (two detail implementations?).
+//   You can see current problems for example in the old NotificationPreferencesFragment.onCreatePreferences()
+//   which only would use pull notifications if the notifications option is enabled.
+class PushNotificationManager @Inject constructor(
+    private val mastodonApi: MastodonApi,
+    private val accountManager: AccountManager,
+    private val sharedPreferences: SharedPreferences,
+    private val context: Context
+): Preference.SummaryProvider<Preference> {
+
+    companion object {
+        const val TAG = "PushNotificationManager"
+        private const val KEY_PUSH_MIGRATION_NOTICE_DISMISSED = "migration_notice_dismissed"
+    }
+
+    // TODO? must be changed/extended when distributors are installed or uninstalled on-the-fly?
+    //   Or there must be an "restart app fully" possibility.
+    private val distributors: List<String> = UnifiedPush.getDistributors(context)
+
+    private fun isUnifiedPushAvailable(): Boolean {
+        return distributors.isNotEmpty()
+    }
+
+    // TODO! there should be an actual decision (possibility) to say "I don't want to use push notifications for Tusky".
+
+    fun canEnablePushNotifications(): Boolean =
+        isUnifiedPushAvailable() && !anyAccountNeedsMigration()
+
+    fun hasPushNotificationsEnabled(account: AccountEntity): Boolean =
+        isUnifiedPushAvailable() && !account.unifiedDistributorName.isNullOrEmpty()
+
+    suspend fun enablePushNotifications() {
+        if (!canEnablePushNotifications()) {
+            return
+        }
+
+        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+
+        accountManager.accounts.forEach {
+            val notificationGroupEnabled = Build.VERSION.SDK_INT < 28 ||
+                nm.getNotificationChannelGroup(it.identifier)?.isBlocked == false
+            val shouldEnable = it.notificationsEnabled && notificationGroupEnabled
+
+            if (shouldEnable) {
+                enableUnifiedPushNotificationsForAccount(it)
+            } else {
+                disableUnifiedPushNotificationsForAccount(it)
+            }
```

Comment:
Why is a method called `enablePushNotifications` possibly disabling push notifications?

Reply:
(it's moved old code)

Could be called `doYourStuff` or `manage` instead - as it is in `PushNotificationManager`.

Maybe `enableOrDisablePerAccount`?